README.md
# MyLittleGallery

A PHP class and templates to create a quick drop-in HTML gallery.

## Throubleshooting the demo script

### Unable to read files with non-ASCII names
#### PHP before 7.1
Check that encoding of `config.php` file itself matches value of GALLERY_FS_ENCODING constant.
#### PHP 7.1+
`config.php` MUST be in `UTF-8`.
For PHP 7.1 and up GALLERY_FS_ENCODING and `$fsEncoding` parameter of the constructor are ignored.

Starting from PHP 7.1, [PHP uses internal_encoding to transcode file names](https://github.com/php/php-src/blob/e33ec61f9c1baa73bfe1b03b8c48a824ab2a867e/UPGRADING#L418).
Before that, file IO under Windows (notably) done using "default" (so-called "ANSI") character set (i.e. CP1251 for Russian cyrillic).

## Compiling single file archive

```
pharcommand.phar pack -f mlg.phar -s stub.php -c bzip2 index.php Gallery.php
```

Afterwards, the resulting `mlg.phar` can be used as index.php in your web gallery directory.
You may want to place config.php next to it to configure the gallery.
By default, it will list all recognized image files in current directory using UTF-8 encoding.


CHANGES
------------------------------------------------------------------------
r987 | anrdaemon | 2019-02-24 22:38:16 +0300 (Sun, 24 Feb 2019) | 2 lines

* Unnecessary whitespace changes.

------------------------------------------------------------------------
r986 | anrdaemon | 2019-02-24 22:37:38 +0300 (Sun, 24 Feb 2019) | 3 lines

* Sorted out usage of iterators' classes.
* Implement \IteratorAggregate instead of direct \Iterator.

------------------------------------------------------------------------
r985 | anrdaemon | 2019-02-24 22:35:43 +0300 (Sun, 24 Feb 2019) | 2 lines

+ Explicitly list extensions used in the project.

------------------------------------------------------------------------
r984 | anrdaemon | 2019-02-24 22:26:12 +0300 (Sun, 24 Feb 2019) | 2 lines

* Include correct configuration file regardless if running in PHP or PHAR mode.

------------------------------------------------------------------------
r983 | anrdaemon | 2019-02-24 22:24:45 +0300 (Sun, 24 Feb 2019) | 2 lines

+ Improve stub.php handling in Makefile.

------------------------------------------------------------------------
r982 | anrdaemon | 2019-02-24 21:59:40 +0300 (Sun, 24 Feb 2019) | 2 lines

+ Sanitize config defaults.

------------------------------------------------------------------------
r766 | anrdaemon | 2018-03-17 15:40:36 +0300 (Sat, 17 Mar 2018) | 2 lines

+ Release building makefile.

------------------------------------------------------------------------


index.sample.php
<?php

if(is_file(__DIR__ . '/Gallery.php'))
{
  require_once __DIR__ . '/Gallery.php';
}

if(!defined('GALLERY_BASE_DIR'))
{
  $confFile = dirname(\Phar::running(false) ?: __FILE__) . '/config.php';
  if(file_exists($confFile))
    require_once $confFile;
}

ini_set('default_charset', 'UTF-8');
setlocale(LC_ALL, 'C.' . GALLERY_FS_ENCODING);

if('cli' === PHP_SAPI)
{
  $_REQUEST['console'] = true;

  if($argc > 1)
  {
    $_REQUEST['preview'] = $argv[1];
  }
  else
  {
    die("Not enough arguments!\n");
  }
}

set_exception_handler(function($e)
{
  static $map = array(
    400 => 'Bad request',
    403 => 'Forbidden',
    404 => 'Not found',
    405 => 'Method not allowed',
    415 => 'Unsupported media type',
    500 => 'Internal server error',
    501 => 'Not implemented',
  );
  $code = isset($map[$e->getCode()]) ? $e->getCode() : 500;
  http_response_code($code);
  exit("<!DOCTYPE html>
<html><head>
<title>{$code} {$map[$code]}</title>
</head><body><h1>{$map[$code]}</h1>
<p>" . htmlspecialchars($e->getMessage()) . "</p>
<hr/>
{$_SERVER['SERVER_SIGNATURE']}</body></html>");
});

if(defined('GALLERY_DESC_FILE'))
{
  $gallery = AnrDaemon\MyLittleGallery\Gallery::fromListfile(new SplFileInfo(GALLERY_BASE_DIR . "/" . GALLERY_DESC_FILE),
    GALLERY_DESC_ENCODING, GALLERY_FS_ENCODING);
}
else
{
  $gallery = AnrDaemon\MyLittleGallery\Gallery::fromDirectory(new SplFileInfo(GALLERY_BASE_DIR),
    null, GALLERY_FS_ENCODING);
}

if(defined('GALLERY_SENDFILE_HEADER'))
{
  $gallery->allowSendFile($gallery->getPrefix('index'), GALLERY_SENDFILE_HEADER);
}

$gallery->setPreviewSize(GALLERY_PREVIEW_X, GALLERY_PREVIEW_Y);

switch(true)
{
  case isset($_REQUEST['preview']):
    $name = basename($_REQUEST['preview']);
    if(!isset($gallery[$name]))
      throw new Exception('No referenced image found.', 404);

    if(!$gallery->thumbnailImage($name))
      throw new Exception("No thumbnail image for '$name'.", 404);

    if(isset($_REQUEST['console']))
    {
      die("Done.\n");
    }

    if($gallery->sendFile("/.preview/$name"))
      break;

    header('Content-type: ' . $gallery[$name]['mime']);
    readfile($gallery->getPath("/.preview/$name", true));

    break;
  case isset($_REQUEST['view']):
    $name = basename($_REQUEST['view']);
    if(!isset($gallery[$name]))
      throw new Exception("Image '$name' not found.", 404);

    if($gallery->sendFile("/$name"))
      break;

    header('Content-type: ' . $gallery[$name]['mime']);
    readfile($gallery->getPath("/$name", true));

    break;
  case isset($_REQUEST['show']):
    $name = basename($_REQUEST['show']);
?><?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE html>
<html>
<head>
<title><?=htmlspecialchars($gallery[$name]['desc'])?> - My Little Gallery</title>
<link rel="INDEX UP" href="<?=htmlspecialchars($gallery->getUrl('index'))?>"/>
<?php if(isset($gallery[$name]['prev'])):?>
<link rel="PREVIOUS" href="<?=htmlspecialchars($gallery->getUrl('view', $gallery[$name]['prev']))?>"/>
<?php endif;
if(isset($gallery[$name]['next'])):?>
<link rel="NEXT" href="<?=htmlspecialchars($gallery->getUrl('view', $gallery[$name]['next']))?>"/>
<?php endif; ?>
<style type="text/css"><!--
h1 {
  padding-top: 1ex;
}

h1, p {
  text-align: center;
}

img {
  max-width: 100%;
  margin: 5px 0px;
  cursor: pointer;
}

div.navFloater {
  position: absolute;
  top: 0px;
  padding: 1em;
}

div.prev {
  left: 0px;
}

div.next {
  right: 0px;
}
//--></style>
</head>
<body>
<h1><?=htmlspecialchars($gallery[$name]['desc'])?></h1>
<p><img src="<?=htmlspecialchars($gallery->getUrl('image', $name))?>
" alt="<?=htmlspecialchars($gallery[$name]['desc'] . ' (' . $gallery->imageFileSize($name, 1024) . "\xC2\xA0kB)")?>
" title="" onclick="window.close();"/></p>
<?php if(isset($gallery[$name]['prev'])):?>
<div class="navFloater prev"><a href="<?=htmlspecialchars($gallery->getUrl('view', $gallery[$name]['prev']))?>
" title="<?=htmlspecialchars($gallery[$gallery[$name]['prev']]['desc'])?>
"><?=htmlspecialchars("<< {$gallery[$gallery[$name]['prev']]['desc']}")?></a></div>
<?php endif;
if(isset($gallery[$name]['next'])):?>
<div class="navFloater next"><a href="<?=htmlspecialchars($gallery->getUrl('view', $gallery[$name]['next']))?>
" title="<?=htmlspecialchars($gallery[$gallery[$name]['next']]['desc'])?>
"><?=htmlspecialchars("{$gallery[$gallery[$name]['next']]['desc']} >>")?></a></div>
<?php endif; ?>
<!-- Forum embed code
[url=<?="{$_SERVER['REQUEST_SCHEME']}://{$_SERVER['SERVER_NAME']}" . htmlspecialchars($gallery->getUrl('view', $name))?>
][img]<?="{$_SERVER['REQUEST_SCHEME']}://{$_SERVER['SERVER_NAME']}" . htmlspecialchars($gallery->getUrl('thumbnail', $name))?>[/img][/url]
[url=<?="{$_SERVER['REQUEST_SCHEME']}://{$_SERVER['SERVER_NAME']}" . htmlspecialchars($gallery->getUrl('view', $name))?>
]<?=htmlspecialchars($gallery[$name]['desc'] . ' (' . $gallery->imageFileSize($name, 1024) . "\xC2\xA0kB)")?>[/url]
-->
</body>
</html>
<?php
    break;
  default:
?><?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE html>
<html>
<head>
<title>My Little Gallery</title>
<style type="text/css"><!--
body {
  max-width: <?=(GALLERY_PREVIEW_X * GALLERY_PREVIEW_WIDTH_FACTOR * GALLERY_COLUMNS)?>px;
  text-align: center;
  margin: 0px auto;
}
div.MyLittleGallery {
  position: relative;
  text-align: left;
}
div.MyLittleGallery div {
  position: relative;
  width: <?=(GALLERY_PREVIEW_X * GALLERY_PREVIEW_WIDTH_FACTOR)?>px;
  margin: 1em 0px;
  display: inline-block;
  text-align: center;
}
div.MyLittleGallery img {
  margin: 1em 0px 0px;
}

//--></style>
</head>
<body><p><a href="./..">Go up</a></p>
<div class="MyLittleGallery"><?=$gallery->showIndex()?></div>
<!-- pre style="text-align: left;"><?php//=print_r($gallery, true)?></pre -->
</body></html>
<?php
}


config.sample.php
<?php

define('GALLERY_BASE_DIR', __DIR__);

/** Your filesystem encoding.
*
* UTF-8 is typical for modern *NIX,
* CP1251 is Russian Windows ANSI codepage.
*/
define('GALLERY_FS_ENCODING', 'UTF-8');
//define('GALLERY_FS_ENCODING', 'CP1251');

/** Enable files' descriptions support.
*
* Otherwise - use direct directory listing by mask.
* Files.bbs/Descript.ion is a popular file description list formats.
*/
//define('GALLERY_DESC_FILE', 'Files.bbs');
//define('GALLERY_DESC_FILE', 'Descript.ion');

/** Specify gallery descriptions encoding.
*
* Only used with GALLERY_DESC_FILE
*
* CP866 for cyrillic descriptions is common.
* UTF-8 is a sane default suitable for drop-in gallery.
*/
//define('GALLERY_DESC_ENCODING', 'CP866');
//define('GALLERY_DESC_ENCODING', 'UTF-8');

/** Define general gallery layout
*/
define('GALLERY_PREVIEW_X', 160);
define('GALLERY_PREVIEW_Y', 120);
define('GALLERY_PREVIEW_WIDTH_FACTOR', 2);
define('GALLERY_COLUMNS', 3);

/** Allow X-SendFile/X-Accel-Redirect
*
* Uncomment one of the following defines, if you know what you are doing.
*
* NB: SERVER CONFIGURATION IS REQUIRED BEFORE ACTIVATING THIS FEATURE!
*/
// Lighttpd, Apache mod_xsendfile
//define('GALLERY_SENDFILE_HEADER', 'X-SendFile');
// nginx
//define('GALLERY_SENDFILE_HEADER', 'X-Accel-Redirect');


Gallery.php
<?php
/** My Little Gallery
*
* A simple drop-in file-based HTML gallery.
*
* $Id: Gallery.php 987 2019-02-24 19:38:16Z anrdaemon $
*/

namespace AnrDaemon\MyLittleGallery;

class Gallery
implements \ArrayAccess, \Countable, \IteratorAggregate
{
  const previewTemplate =
    '<div><a href="%1$s" target="_blank"><img src="%2$s" alt="%3$s"/></a><p><a href="%1$s" target="_blank">%3$s</a></p></div>';

  const defaultTypes = 'gif|jpeg|jpg|png|wbmp|webp';

  // All paths are UTF-8! (Except those from \SplFileInfo)
  protected $path; // Gallery base path
  protected $prefix = array(); // Various prefixes for correct links construction
  protected $params = array(); // Files list
  protected $extensions = array(); // Allowed extensions

  // FS encoding
  protected $cs;

  // Numbers formatter
  protected $nf;

  // Preview settings
  protected $pWidth;
  protected $pHeight;
  protected $template;

  // X-SendFile settings
  protected $sfPrefix;
  protected $sfHeader = 'X-SendFile';

  protected function fromFileList(array $list)
  {
    $prev = null;
    foreach($list as $fname)
    {
      if(is_dir($fname))
        continue;

      $name = iconv($this->cs, 'UTF-8', basename($fname));
      $this->isSaneName($name);

      $meta = getimagesize($fname);
      if($meta === false || $meta[0] === 0 || $meta[1] === 0)
        continue;

      $this->params[$name] = new \ArrayObject([
        'desc' => $name,
        'path' => "{$this->path}/{$name}",
        'width' => $meta[0],
        'height' => $meta[1],
        'mime' => $meta['mime'],
      ]);
      if(isset($prev))
      {
        $this->params[$name]['prev'] = $prev;
        $this->params[$prev]['next'] = $name;
      }
      $prev = $name;
    }

    return $this;
  }

  public static function fromListfile(\SplFileInfo $target, $charset = 'CP866', $fsEncoding = null)
  {
    $path = $target->getRealPath();

    if(empty($path))
      throw new \Exception('Can\'t use empty path.', 500);

    if(is_dir($path))
      throw new \Exception('Target is a directory', 500);

    $self = new static($target->getPathInfo(), null, $fsEncoding);

    $f = iconv($charset, 'UTF-8', file_get_contents($path));

    if(preg_match_all('/^(\"?)(?P<name>[^\"]+?)\1\s+(?P<desc>.*?)\s*$/um', $f, $ta, PREG_SET_ORDER))
    {
      $prev = null;
      foreach($ta as $a)
      {
        $name = basename(trim($a['name']));
        $self->isSaneName($name);

        $meta = getimagesize(iconv('UTF-8', $self->cs, "{$self->path}/$name"));
        if($meta === false || $meta[0] === 0 || $meta[1] === 0)
          continue;

        $self->params[$name] = new \ArrayObject([
          'desc' => $a['desc'],
          'path' => "{$self->path}/{$name}",
          'width' => $meta[0],
          'height' => $meta[1],
          'mime' => $meta['mime'],
        ]);
        if(isset($prev))
        {
          $self->params[$name]['prev'] = $prev;
          $self->params[$prev]['next'] = $name;
        }
        $prev = $name;
      }
    }

    return $self;
  }

  public static function fromDirectory(\SplFileInfo $target, array $extensions = null, $fsEncoding = null)
  {
    if(empty($extensions))
    {
      $extensions = explode('|', static::defaultTypes);
    }

    $mask = "*.{" . implode(',', $extensions) . "}";

    return static::fromCustomMask($target, $mask, $fsEncoding);
  }

  public static function fromCustomMask(\SplFileInfo $target, $mask, $fsEncoding = null)
  {
    $path = $target->getRealPath();

    if(empty($path))
      throw new \Exception('Can\'t use empty path.', 500);

    if(!is_dir($path))
      throw new \Exception('Target is not a directory', 500);

    $self = new static($target, null, $fsEncoding);

    return $self->fromFileList(glob("{$path}/" . iconv('UTF-8', $self->cs, $mask), GLOB_BRACE | GLOB_MARK));
  }
  /**
  * $template($show, $preview, $description)
  */
  public function showIndex($template = null)
  {
    if(empty($template))
    {
      $template = $this->template;
    }

    $gp = '';
    foreach($this->params as $f => $d)
    {
      $gp .= sprintf($template, htmlspecialchars($this->prefix['view'] . rawurlencode($f)),
        htmlspecialchars($this->prefix['thumbnail'] . rawurlencode($f)),
        htmlspecialchars($d['desc'] . ' (' . $this->imageFileSize($f, 1024) . "\xC2\xA0kB)"));
    }

    return $gp;
  }

  public function setNumberFormatter($locale = 'en_US.UTF-8', $style = \NumberFormatter::DECIMAL, $pattern = '')
  {
    $this->nf = new \NumberFormatter($locale, $style, $pattern);

    return $this;
  }

  public function allowSendFile($prefix = null, $header = null)
  {
    $this->sfPrefix = $prefix;
    $this->sfHeader = trim($header) ?: 'X-SendFile';

    return $this;
  }

  public function sendFile($path)
  {
    if(!isset($this->sfPrefix))
      return false;

    header_register_callback(
      function()
      {
        /*
        Accept-Ranges
        Cache-Control
        Content-Disposition
        Content-Type
        Expires
        Set-Cookie
        */
        header_remove('Accept-Ranges');
        header_remove('Content-Type');
      }
    );

    header("{$this->sfHeader}: {$this->sfPrefix}" . urlencode("$path"));

    return true;
  }

  public function imageFileSize($name, $divisor = 1)
  {
    return $this->nf->format(ceil(filesize(iconv('UTF-8', $this->cs, "{$this->path}/$name")) / $divisor));
  }

  public function imagePreviewExists($name)
  {
    if(isset($this->params[$name]['preview']))
      return !empty($this->params[$name]['preview']);

    $fname = iconv('UTF-8', $this->cs, "{$this->path}/.preview/$name");
    return $this->params[$name]['preview'] = file_exists($fname);
  }

  public function setPreviewSize($width = null, $height = null)
  {
    if((int)$width < 0 || (int)$height < 0)
      throw new \Exception('Thumbnail dimensions can\'t be negative.', 500);

    $this->pWidth = (int)$width ?: 160;
    $this->pHeight = (int)$height ?: 120;

    return $this;
  }

  public function setPrefix($name, $prefix)
  {
    if(!isset($this->prefix[$name]))
      throw new \Exception("Unknown prefix '$name'.", 500);

    $this->prefix[$name] = $prefix;

    return $this;
  }

  public function setTemplate($template = null)
  {
    $this->template = empty($template)
      ? static::previewTemplate
      : $template;

    return $this;
  }

  public function getPrefix($name)
  {
    return $this->prefix[$name];
  }

  public function getPath($name = null, $local = null)
  {
    $path = $this->path;
    if(isset($name))
    {
      $path .= $name;
    }

    if($local)
    {
      $path = iconv('UTF-8', $this->cs, $path);
    }

    return $path;
  }

  public function getUrl($prefix, $name = '')
  {
    return $this->prefix[$prefix] . ($prefix === 'index' ? '/' : rawurlencode($name));
  }

  public function thumbnailImage($name)
  {
    static $gdSave = array(
      'image/gif' => 'imagegif',
      'image/jpeg' => 'imagejpeg',
      'image/png' => 'imagepng',
      'image/vnd.wap.wbmp' => 'imagewbmp',
      'image/webp' => 'imagewebp',
    );

    if(!isset($this->params[$name]))
      throw new \Exception("Image '$name' is not registered in the gallery.", 404);

    $path = $this->getPath("/.preview/$name", true);
    if(is_file($path))
      return true;

    try
    {
      set_error_handler(function($s, $m, $f, $l, $c = null) { throw new \ErrorException($m, 0, $s, $f, $l); });

      if(class_exists('Imagick'))
      {
        $img = new \Imagick("{$this->path}/$name");
        $img->thumbnailImage($this->pWidth, $this->pHeight, true);
        $img->writeImage("{$this->path}/.preview/$name");
      }
      elseif(function_exists('imagecreatefromstring'))
      {
        $src = imagecreatefromstring(file_get_contents($this->getPath("/$name", true)));
        if($src === false)
          throw new \Exception("The file '$name' can't be interpreted as image.", 500);

        $oFactor = $this->params[$name]['width'] / $this->params[$name]['height'];
        $tFactor = $this->pWidth / $this->pHeight;
        if($oFactor >= $tFactor)
        {
          $w = $this->pWidth;
          $h = min($this->pHeight, ceil($this->pWidth / $oFactor));
        }
        else
        {
          $w = min($this->pWidth, ceil($this->pHeight * $oFactor));
          $h = $this->pHeight;
        }

        $img = imagecreatetruecolor($w, $h);
        if(!imagecopyresampled($img, $src, 0, 0, 0, 0, $w, $h, $this->params[$name]['width'], $this->params[$name]['height']))
          throw new \Exception("Unable to create thumbnail for '$name'.", 500);

        $gdSave[$this->params[$name]['mime']]($img, $path);
      }
      else
        throw new \ErrorException('Imagick or gd2 extension is required to create thumbnails at runtime.', 501);

      restore_error_handler();
    }
    catch(\Exception $e)
    {
      restore_error_handler();
      if(!is_dir(dirname($path)))
      {
        mkdir(dirname($path));
        return false;
      }

      throw $e;
    }

    return true;
  }

  // FIX anti-exploit
  public function isSaneName($fname)
  {
    $name = basename($fname);
    if(preg_match('/[^!#$%&\'()+,\-.;=@\[\]^_`{}~\p{L}\d\s]/uiS', $name))
      throw new \Exception("Invalid character in name '$name'.", 400);

    if(!preg_match('{.+\.(' . implode('|', array_map('preg_quote', $this->extensions)) . ')$}ui', $name))
      throw new \Exception('Invalid filename extension.', 400);

    return true;
  }

// Magic!

  protected function __construct(\SplFileInfo $path, array $extensions = null, $fsEncoding = null)
  {
    if(version_compare(PHP_VERSION, '7.1', '<'))
    {
      $this->cs = trim($fsEncoding) ?: 'UTF-8';
    }
    else
    {
      ini_set('internal_encoding', 'UTF-8');
      $this->cs = 'UTF-8';
    }

    $this->path = iconv($this->cs, 'UTF-8', realpath($path->getRealPath()));

    // $path is not necessarily equals $path->getRealPath()
    // Work off original $path
    $this->prefix['index'] = iconv($this->cs, 'UTF-8', substr(realpath(realpath($path)), strlen(realpath(realpath($_SERVER['DOCUMENT_ROOT'])))));
    $this->prefix['view'] = $this->prefix['index'] . '/?show=';
    $this->prefix['thumbnail'] = $this->prefix['index'] . '/?preview=';
    $this->prefix['image'] = $this->prefix['index'] . '/?view=';

    $this->setNumberFormatter();
    $this->setPreviewSize();
    $this->setTemplate();

    if(empty($extensions))
    {
      $this->extensions = explode('|', static::defaultTypes);
    }
    else
    {
      $masks = array();
      foreach(array_map('trim', $extensions) as $ext)
      {
        if(empty($ext))
          continue;

        $masks[] = $ext;
      }

      if(empty($masks))
        throw new \Exception('File extensions can\'t be empty strings.', 500);

      $this->extensions = $masks;
    }
  }

// ArrayAccess

  public function offsetSet($offset, $value)
  {
    $this->params[$offset] = $value;
  }

  public function offsetGet($offset)
  {
    return $this->params[$offset];
  }

  public function offsetExists($offset)
  {
    return isset($this->params[$offset]);
  }

  public function offsetUnset($offset)
  {
    unset($this->params[$offset]);
  }

// Countable

  public function count()
  {
    return count($this->params);
  }

// IteratorAggregate

  public function getIterator()
  {
    return new \ArrayIterator($this->params);
  }
}
